### PR TITLE
Document //flutter/shell/common/rasterizer

### PR DESCRIPTION
**Pull request by chinmaygarde**

### CODE REVIEW

File: shell/common/rasterizer.h [hunk 1]
```diff
@@ -20,78 +20,367 @@
...
 
-/// Takes |LayerTree|s and draws its contents.
+//------------------------------------------------------------------------------
+/// The rasterizer is a component owned by the shell that resides on the GPU
+/// task runner. Each shell owns exactly one instance of a rasterizer. The
+/// rasterizer may only be created, used and collected on the GPU task runner.
+///
+/// The rasterizer owns the instance of the currently active on-screen render
+/// surface. On this surface, it renders the contents of layer trees submitted
+/// to it by the engine (which lives on the UI task runner).
```
Comment (liyuqian) [inline shell/common/rasterizer.h]: Maybe explicitly specify `flutter::Engine` so readers won't be confused with the engine as a whole (e.g., flutter/engine repo).

File: shell/common/rasterizer.h [hunk 2]
```diff
@@ -20,78 +20,367 @@
...
+  //----------------------------------------------------------------------------
+  /// @brief      Rasterizers may be created well before an on-screen surface is
+  ///             available for rendering. Shells usually create a rasterizer in
+  ///             their constructors. Once an on-screen surface is available
+  ///             however, one may be provided to the rasterizer using this
+  ///             call. No rendering may occur before this call. The surface is
+  ///             held till the balancing call to `Rasterizer::Teardown` is
+  ///             made. Calling a setup before tearing down the previous surface
+  ///             (if this is not the first time the surface has been setup) is
+  ///             user error.
```
Comment (liyuqian) [inline shell/common/rasterizer.h]: "is A user error"?

File: shell/common/rasterizer.h [hunk 3]
```diff
@@ -20,78 +20,367 @@
...
+  ///
+  /// @param[in]  surface  The on-screen render surface.
+  ///
   void Setup(std::unique_ptr<Surface> surface);
 
+  //----------------------------------------------------------------------------
+  /// @brief      Releases the previously setup on-screen render surface and
+  ///             collects associated resources. No more rendering may occur
+  ///             till the next call to `Rasterizer::Setup` with a new render
+  ///             surface. Calling a teardown without a setup is user error.
```
Comment (liyuqian) [inline shell/common/rasterizer.h]: "is A user error"?

File: shell/common/rasterizer.h [hunk 4]
```diff
@@ -20,78 +20,367 @@
...
-  // This method must be called from the GPU task runner.
+  //----------------------------------------------------------------------------
+  /// @brief      Notifies the rasterizer that there is a low memory situation
+  ///             and it must purge as many unnecessary resources as possible.
+  ///             Currently, the Skia context associated with onscreen rendering
+  ///             is told to free GPU resources.
+  ///
   void NotifyLowMemoryWarning() const;
 
+  //----------------------------------------------------------------------------
```
Comment (liyuqian) [inline shell/common/rasterizer.h]: Maybe add "The weak pointer can be copied to other threads, but the rasterizer it references to may only be accessed..."

File: shell/common/rasterizer.h [hunk 5]
```diff
@@ -20,78 +20,367 @@
...
+  /// @brief      Gets a weak pointer to the rasterizer. The rasterizer may only
+  ///             be accessed on the GPU task runner.
+  ///
+  /// @return     The weak pointer to the rasterizer.
+  ///
   fml::WeakPtr<Rasterizer> GetWeakPtr() const;
 
+  //----------------------------------------------------------------------------
+  /// @brief      Gets a weak pointer to the rasterizer. The rasterizer is also
```
Comment (liyuqian) [inline shell/common/rasterizer.h]: I guess this is no longer needed if https://github.com/flutter/engine/pull/9813 lands first?

File: shell/common/rasterizer.h [hunk 6]
```diff
@@ -20,78 +20,367 @@
...
+  ///
   fml::WeakPtr<SnapshotDelegate> GetSnapshotDelegate() const;
 
+  //----------------------------------------------------------------------------
+  /// @brief      Sometimes, it may be necessary to render the same frame again
+  ///             without having to wait for the framework to build a whole new
+  ///             layer tree describing the same contents. One such case is when
+  ///             external textures (video or camera streams for example) are
+  ///             updated in an otherwise static layer tree. To support this use
+  ///             case, the rasterizer holds onto the last rendered layer tree.
```
Comment (liyuqian) [inline shell/common/rasterizer.h]: Maybe leave a link to https://github.com/flutter/flutter/issues/33939 to clarify that we intend to reduce such redundant rendering.

File: shell/common/rasterizer.h [hunk 7]
```diff
@@ -20,78 +20,367 @@
...
+  ///             them from one frame to the next. Using this call, embedders
+  ///             may set the maximum bytes cached by Skia in its caches
+  ///             dedicated to on-screen rendering.
+  ///
+  /// @attention  This cache setting will be invalidated when the surface is
+  ///             torn down via `Rasterizer::Teardown`. This call must be made
+  ///             again with new limits after surface re-acquisition.
+  ///
+  /// @attention  This cache does not describe the entirety of GPU resources
+  ///             that may be cached. The raster cache also holds very large GPU
```
Comment (liyuqian) [inline shell/common/rasterizer.h]: Maybe add a reference to `|RasterCache|` (so they know that raster cache is not just a concept, but an actual class)?

File: shell/common/rasterizer.h [hunk 8]
```diff
@@ -20,78 +20,367 @@
...
+  ///
   flutter::TextureRegistry* GetTextureRegistry();
 
+  //----------------------------------------------------------------------------
+  /// @brief      Takes the next item from the layer tree pipeline and executes
+  ///             the GPU thread frame workload for that pipeline item to render
+  ///             a frame on the on-screen surface.
+  ///
+  ///             Why does the draw call take a layer tree pipeline and not the
+  ///             layer tree directly?
```
Comment (liyuqian) [inline shell/common/rasterizer.h]: Maybe add a note like "|Rasterizer::DoDraw| takes a layer tree and rasterizes it; The |Rasterizer::Draw| takes the pipeline as it needs to do additional pipeline maintenance work".

File: shell/common/rasterizer.h (hunk 1 above)
Comment (chinmaygarde) [reply shell/common/rasterizer.h]: Done.

File: shell/common/rasterizer.h [hunk 9]
```diff
@@ -20,78 +20,367 @@
...
+/// The rasterizer is a component owned by the shell that resides on the GPU
+/// task runner. Each shell owns exactly one instance of a rasterizer. The
+/// rasterizer may only be created, used and collected on the GPU task runner.
+///
+/// The rasterizer owns the instance of the currently active on-screen render
+/// surface. On this surface, it renders the contents of layer trees submitted
+/// to it by the engine (which lives on the UI task runner).
+///
+/// The primary components owned by the rasterizer are the compositor context
+/// and the on-screen render surface. The compositor context all the GPU state
```
[the remark this answers is not in the document]
Comment (chinmaygarde) [reply shell/common/rasterizer.h]: Done.

File: shell/common/rasterizer.h (hunk 2 above)
Comment (chinmaygarde) [reply shell/common/rasterizer.h]: I think this usage is correct.

File: shell/common/rasterizer.h (hunk 3 above)
Comment (chinmaygarde) [reply shell/common/rasterizer.h]: I think this usage is correct.

File: shell/common/rasterizer.h (hunk 4 above)
Comment (chinmaygarde) [reply shell/common/rasterizer.h]: Isn't that a property of all weak pointers? That is documented in FML. This weak pointer is not special in any way.

File: shell/common/rasterizer.h (hunk 5 above)
Comment (chinmaygarde) [reply shell/common/rasterizer.h]: I'll remove it then.

File: shell/common/rasterizer.h (hunk 6 above)
Comment (chinmaygarde) [reply shell/common/rasterizer.h]: Done. I added an `@bug` line.

File: shell/common/rasterizer.h (hunk 8 above)
Comment (chinmaygarde) [reply shell/common/rasterizer.h]: Done.

File: shell/common/rasterizer.h (hunk 7 above)
Comment (chinmaygarde) [reply shell/common/rasterizer.h]: Done.

**Comment by chinmaygarde** [comment]
Rebased on ToT.

**Pull request merged**